### PR TITLE
Authorization feature

### PR DESCRIPTION
**Pull request by alexkochnev1987**
1. реализована авторизация и регистрация пользователя. Добавлена валидация данных. Добавлены интерфейсы. 
2. 
image

### CODE REVIEW

**Comment by malickalex1975** [comment]
well done

**Review by malickalex1975 — APPROVED** [review]

File: rs-lang-app/src/app/authorization/login/login.component.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  username = '';
+  form: FormGroup;
+  constructor(private auth: AuthService) {
+    this.form = new FormGroup({});
+  }
+
+  ngOnInit(): void {
+    this.switchLogin();
+  }
+  switchRegister() {
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.ts:21]: Blank space between methods, please.

File: rs-lang-app/src/app/core/services/auth.service.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  providedIn: 'root',
+})
+export class AuthService {
+  private token = '';
+  constructor(private http: HttpClient) {}
+  login(user: User): Observable<LoginResponse> {
+    return this.http
+      .post<LoginResponse>(`${url + QueryParams.logIn}`, {
+        email: user.email,
+        password: user.password,
```
Comment (TELEUZI) [inline rs-lang-app/src/app/core/services/auth.service.ts:22]: Use object destructuring, please.

File: rs-lang-app/src/app/core/services/auth.service.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      })
+      .pipe(
+        tap(o => {
```
Comment (TELEUZI) [inline rs-lang-app/src/app/core/services/auth.service.ts:25]: Use name more readable than 'o', please.

File: rs-lang-app/src/app/core/services/auth.service.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+          this.setToken(o.token);
+          this.setLocalStorage('auth-token', this.token);
```
Comment (TELEUZI) [inline rs-lang-app/src/app/core/services/auth.service.ts:27]: 'auth-token' should be constant in variables file.

File: rs-lang-app/src/app/core/services/auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        })
+      );
+  }
+  setToken(token: string) {
+    this.token = token;
+  }
+  getToken() {
```
Comment (TELEUZI) [inline rs-lang-app/src/app/core/services/auth.service.ts:34]: Use blank space between methods for better readability.

File: rs-lang-app/src/app/authorization/login/login.component.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        console.log(error);
+      },
+    });
+  }
+  onRegistration() {
+    this.form.disable();
+    console.log(this.form.value);
+    this.auth.register(this.form.value).subscribe({
+      next: r => {
+        console.log(r);
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.ts:64]: What will we do on success registration? May be redirect user somewhere?

File: rs-lang-app/src/app/authorization/login/login.component.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+      },
+      error: error => {
+        this.form.enable();
+        console.log(error);
+      },
+    });
+  }
+  onRegistration() {
+    this.form.disable();
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.ts:60]: Why we disable form?

File: rs-lang-app/src/app/authorization/login/login.component.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    console.log(this.form.value);
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.ts:61]: Delete, such logs, please.

File: rs-lang-app/src/app/authorization/login/login.component.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        Validators.required,
+        Validators.minLength(8),
+      ]),
+    });
+  }
+  onSubmit() {
+    this.form.disable();
+    console.log(this.form.value);
+    this.auth.login(this.form.value).subscribe({
+      next: r => {
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.ts:49]: Use better naming than 'r', please.

File: rs-lang-app/src/app/authorization/login/login.component.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+      email: new FormControl(null, [Validators.required, Validators.email]),
+      password: new FormControl(null, [
+        Validators.required,
+        Validators.minLength(8),
+      ]),
+    });
+  }
+  onSubmit() {
+    this.form.disable();
+    console.log(this.form.value);
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.ts:47]: Delete, such logs, please.

File: rs-lang-app/src/app/authorization/login/login.component.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    this.form = new FormGroup({
+      email: new FormControl(null, [Validators.required, Validators.email]),
+      password: new FormControl(null, [
+        Validators.required,
+        Validators.minLength(8),
+      ]),
+    });
+  }
+  onSubmit() {
+    this.form.disable();
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.ts:46]: Why we disable form?

File: rs-lang-app/src/app/authorization/constants.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
+export const url = 'https://learnlangapp1.herokuapp.com';
+export enum QueryParams {
+  logIn = '/signin',
+  register = '/users',
+  words = '/words',
+}
+export interface LoginResponse {
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/constants.ts:7]: ???
Why do we have 'string' as value, not as type?

File: rs-lang-app/src/app/authorization/constants.ts [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+  message: 'string';
+  token: 'string';
+  refreshToken: 'string';
+  userId: 'string';
+  name: 'string';
+}
+
+export interface RegisterResponse {
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/constants.ts:15]: ???
Why do we have 'string' as value, not as type?
And it is interface, not constant.

File: rs-lang-app/src/app/authorization/login/login.component.html [hunk 14]
```diff
@@ -0,0 +1,97 @@
+<app-header></app-header>
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.html:1]: If header is the same on every page, we should put it in app.component.html

File: rs-lang-app/src/app/authorization/login/login.component.html [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+<form [formGroup]="form" class="authorization">
+  <h3 *ngIf="!username && login">Log in</h3>
+  <h3 *ngIf="!username && !login">Registration</h3>
+  <h2 *ngIf="username">Hello {{ username }}</h2>
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.html:5]: Divide into login and registration component.

File: rs-lang-app/src/app/authorization/login/login.component.html [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+  </button>
+  <button *ngIf="username" class="logout" (click)="logOut()">Logout</button>
+  <div class="text-image">
+    <div class="text-wrapper">
+      <div class="text-done">
+        <img src="../../../assets/img/Vector 11.svg" alt="done" />
+        <p>Save hard words</p>
+      </div>
+      <div class="text-done">
+        <img src="../../../assets/img/Vector 11.svg" alt="done" />
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.html:84]: Just use  `done` without relative paths.

File: rs-lang-app/src/app/authorization/login/login.component.html [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+        <p>Save your progress</p>
+      </div>
+      <div class="text-done">
+        <img src="../../../assets/img/Vector 11.svg" alt="done" />
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.html:88]: `assets/img...`

File: rs-lang-app/src/app/authorization/login/login.component.html [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+        <p>Save your statistics</p>
+      </div>
+    </div>
+    <div class="wrapper-image">
+      <img src="../../../assets/img/006 1.png" alt="book" />
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.html:93]: Same here.

File: rs-lang-app/src/app/authorization/login/login.component.scss [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+    text-align: center;
+  }
+}
+.text-image{
+  display: flex;
+  justify-content: center;
+  gap: 30px;
+}
+.text-wrapper{
+display: flex;
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.scss:18]: Fix prettier, please.

File: rs-lang-app/src/app/authorization/login/login.component.scss [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+}
+.input {
+  margin-bottom: 10px;
+  width: 332px;
+  height: 34px;
+  border-radius: 10px;
+  border: none;
+}
+button{
+  // border: 1px dashed #ffffff;
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.scss:47]: No commented styles, please.

File: rs-lang-app/src/app/authorization/login/login.component.scss [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+  line-height: 20px;
+}
+.error{
+  -webkit-user-select: none;
+  transform: translateY(-8px);
+  margin: 0%;
+  font-family: 'New Rubrik Edge';
+  font-style: normal;
+  font-weight: 400;
+  font-size: 14px;
```
Comment (TELEUZI) [inline rs-lang-app/src/app/authorization/login/login.component.scss:92]: Put font-sizes and colors in sass variables, please.

File: rs-lang-app/src/app/authorization/login/login.component.ts (hunk 7 above)
Comment (alexkochnev1987) [reply rs-lang-app/src/app/authorization/login/login.component.ts]: Чтобы несолько раз не передавались данные на сервер, так только один

File: rs-lang-app/src/app/authorization/login/login.component.ts (hunk 11 above)
Comment (alexkochnev1987) [reply rs-lang-app/src/app/authorization/login/login.component.ts:34]: Чтобы пользователь мог отправить форму только один раз.

**Review by TELEUZI — APPROVED** [review]

**Pull request merged**